### PR TITLE
Add flops profiler tutorial

### PR DESCRIPTION
**Pull request by cli99**
Added the flops profiler tutorial, configuration, and feature to the website. Also fixed names of some flops profiler function parameters.

### CODE REVIEW

File: deepspeed/profiling/flops_profiler/README.md [hunk 1]
```diff
@@ -35,216 +38,324 @@ Top 3 modules in flops at depth 2 are {'Conv2d': '421.91 MMACs', 'Linear': '11.1
...
+        38.89 M, 100.00% Params, 103.62 GMACs, 100.00% MACs, 4.73 ms, 14.21% time, 4.4e+01 TFLOPS,
+        (embedding): Embedding(
+          26.28 M, 67.57% Params, 0 MACs, 0.00% MACs, 307.56 us, 0.92% time, 0.0 TFLOPS,
+          (word_embeddings): VocabParallelEmbedding(25.76 M, 66.23% Params, 0 MACs, 0.00% MACs, 130.41 us, 0.39% time, 0.0 TFLOPS, )
+          (position_embeddings): Embedding(524.29 k, 1.35% Params, 0 MACs, 0.00% MACs, 68.19 us, 0.20% time, 0.0 TFLOPS, 1024, 512)
+          (embedding_dropout): Dropout(0, 0.00% Params, 0 MACs, 0.00% MACs, 56.27 us, 0.17% time, 0.0 TFLOPS, p=0.1, inplace=False)
+        )
+        (transformer): ParallelTransformer(
+          12.61 M, 32.43% Params, 103.62 GMACs, 100.00% MACs, 4.4 ms, 13.22% time, 4.7e+01 TFLOPS,
+          (layers): ModuleList(
```
Comment (samyam) [inline deepspeed/profiling/flops_profiler/README.md:96]: Why is time percent and TFlops 0 for ModuleList?

File: deepspeed/profiling/flops_profiler/README.md [hunk 2]
```diff
@@ -35,216 +38,324 @@ Top 3 modules in flops at depth 2 are {'Conv2d': '421.91 MMACs', 'Linear': '11.1
...
-    model = pt_models[args.model]()
-
-    if torch.cuda.is_available():
-        model.cuda(device=args.device)
-
+with torch.cuda.device(0):
+    model = models.alexnet()
     batch_size = 256
-    macs, params, steps = get_model_profile(model, # the PyTorch model to be profiled
+    macs, params = get_model_profile(model=model, # model
```
Comment (samyam) [inline deepspeed/profiling/flops_profiler/README.md:153]: It would be good to how how the  model is called without the profiling so that the input_res and input_consturctors are clearer

File: deepspeed/profiling/flops_profiler/README.md [hunk 3]
```diff
@@ -35,216 +38,324 @@ Top 3 modules in flops at depth 2 are {'Conv2d': '421.91 MMACs', 'Linear': '11.1
...
 
-if __name__ == '__main__':
-    bert_tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
+with torch.cuda.device(0):
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
     model = BertForSequenceClassification.from_pretrained('bert-base-uncased')
-    macs, params, steps = get_model_profile(
+    batch_size = 5
+    seq_len = 128
+    macs, params = get_model_profile(
```
Comment (samyam) [inline deepspeed/profiling/flops_profiler/README.md:234]: It would be good to how how the model is called without the profiling so that the input_res and input_consturctors are clearer. Maybe have:
if profile:
macs, params = get_model_profile
else:
output/loss = model(....)

File: deepspeed/profiling/flops_profiler/README.md [hunk 4]
```diff
@@ -35,216 +38,324 @@ Top 3 modules in flops at depth 2 are {'Conv2d': '421.91 MMACs', 'Linear': '11.1
...
+  * `print_model_profile()` - prints the model graph with the measured profile attached to each module
+  * `print_model_aggregated_profile(module_depth=-1, top_modules=3)` - prints the names of the top modules in terms of aggregated time, flops, and parameters at depth `module_depth`.
+  * `end_profile()` - ends profiling and cleans up. This should be invoked at the end of the profiling and after any printing method.
+
+#### Example Training Workflow
 
-# Output:
-# Number of multiply-adds:        21.74 GMACs
-# Number of parameters:           109.48 M
+Below is an example of this usage in a typical training workflow.
```
Comment (samyam) [inline deepspeed/profiling/flops_profiler/README.md:327]: In this mode, is the profiler capturing only the forward, or forward backward and step? Can we make this more explicit?

File: docs/_tutorials/flops-profiler.md [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+
+## Multi-GPU, Multi-node Runs
+
+For models running on multi-GPU or multi-node, only the model parallelism (e.g. ```--model-parallel-size``` in [Megatron-LM](https://github.com/NVIDIA/Megatron-LM)) affects the number of flops and parameters profiled, i.e.,
+`model_parallel_size * flops = total_flops` and `model_parallel_size * parameters = total_parameters`. The number of GPUs or nodes does not affect the output profile.
+
+
+## Usage With the DeepSpeed Runtime
```
Comment (samyam) [inline docs/_tutorials/flops-profiler.md:58]: Can we have a Usage section with an opening paragraph that describes the following structure.? You already have the structure as below, I think having a paragraph that describes the structure upfront makes it easier for the reader to follow

Usage
- With DeepSpeed Runtime
- Without DeepSpeed Runtime
- ---Inference
- ---Training

-

File: deepspeed/profiling/flops_profiler/README.md [hunk 6]
```diff
@@ -35,216 +38,324 @@ Top 3 modules in flops at depth 2 are {'Conv2d': '421.91 MMACs', 'Linear': '11.1
...
+## Multi-GPU, Multi-node Runs
 
-Refer to the [installaiton of DeepSpeed](https://www.deepspeed.ai/getting-started/#installation) for more information.
+For models running on multi-GPU or multi-node, only the model parallelism (e.g. ```--model-parallel-size``` in [Megatron-LM](https://github.com/NVIDIA/Megatron-LM)) affects the number of flops and parameters profiled, i.e.,
+`model_parallel_size * flops = total_flops` and `model_parallel_size * parameters = total_parameters`. The number of GPUs or nodes does not affect the output profile.
 
-## Usage
 
-### With the DeepSpeed runtime
+## Usage With the DeepSpeed Runtime
```
Comment (samyam) [inline deepspeed/profiling/flops_profiler/README.md:55]: Can we have a Usage section with an opening paragraph that describes the following structure.? You already have the structure as below, I think having a paragraph that describes the structure upfront makes it easier for the reader to follow

Usage

With DeepSpeed Runtime

Without DeepSpeed Runtime

---Inference

---Training

File: deepspeed/profiling/flops_profiler/README.md (hunk 4 above)
Comment (cli99) [reply deepspeed/profiling/flops_profiler/README.md]: The profiler only captures the forward. I clarify this through the README.

File: deepspeed/profiling/flops_profiler/README.md (hunk 3 above)
Comment (cli99) [reply deepspeed/profiling/flops_profiler/README.md]: Made changes as suggested

File: deepspeed/profiling/flops_profiler/README.md (hunk 1 above)
Comment (cli99) [reply deepspeed/profiling/flops_profiler/README.md:113]: Fixed.

File: deepspeed/profiling/flops_profiler/README.md (hunk 6 above)
Comment (cli99) [reply deepspeed/profiling/flops_profiler/README.md]: Made the changes as suggested. I added a table of content for the usage section as well.

**Review by samyam — APPROVED** [review]

**Comment by Nuclear6** [comment]
In the inference module, can I add the performance analysis tutorial of llama2?
